### ISSUE TITLE
Improvement: Use mechanic pathfinding instead of x-y search to find closest mechanic to rides

### ISSUE THREAD

**Issue by bitman2049** [OpenRCT2/OpenRCT2#9351]
When a ride breaks down, find_closest_mechanic only considers the closest mechanic in x-y space, completely ignoring paths. It's good enough for flat parks but fails in vertical or partitioned layouts. A better approach would be to scan through all available mechanics, run pathfind from each mechanic to the ride's exit/entrance, and pick the mechanic with the shortest path.

- [*] Reproducible in RCT2 (vanilla)?

**Steps to reproduce:**

See the attached save file. There are 9 rollercoasters stacked on top of each other, and each has a mechanic patrolling outside the exit station. However, when a ride breaks down, it considers all 9 mechanics to be the same distance even though a specific one is always closest. As a result, mechanics get called to fix rides all over the park instead of working on the one they're patrolling next to, they get stuck in their patrol areas, and the rides never get inspected or fixed.

Removing patrol areas just means that mechanics can reach the rides, eventually, but the response time is beyond unreasonable.

**Save game:**

[Altair Inc. Amusement Division.zip](https://github.com/OpenRCT2/OpenRCT2/files/3245036/Altair.Inc.Amusement.Division.zip)

**Comment by ocalhoun6** [comment]
For a quick and dirty fix that's less demanding on system resources, you could just have it consider the mechanic's location in x, y, _and_ z directions.

That could still have issues with some path layouts, but it could be much more efficient in vertical parks.

**Comment by bitman2049** [comment]
I agree that that's a less demanding fix, but how expensive are pathfinding calls? My understanding is that guests go through pathfinding all the time. Would mechanics add much more to the load?

**Comment by ocalhoun6** [comment]
Oh, and while we're at it, if I remember right, the mechanic's position is determined relative to the ride _entrance_, even though it's the distance to the ride exit that's most important, since that's where the mechanic will actually be going.

That's not a problem usually, but it can be inefficient in some edge cases of weird path layouts, extremely long stations, and/or rides where the entrance and exit are on different station platforms.

**Comment by bitman2049** [comment]
It looks like the game checks for the station's exit first, then checks for its entrance if it doesn't have an exit. See [Ride.cpp, line 2829.](https://github.com/OpenRCT2/OpenRCT2/blob/dd0649dfed9a67e0c2ab330741042ee71ecfc212/src/openrct2/ride/Ride.cpp#L2829)

**Comment by duncanspumpkin** [comment]
The pathfinding code would be hard to use for this. Maybe after the pathfinding code is fixed to be modular then we could do this but until then it would be a pain to do.

**Comment by bitman2049** [comment]
Hmm, OK. Good to know. I've forked this and I want to play with it a bit. In the meantime, XYZ distance should be pretty simple to implement.
